### DISCUSSION TITLE
Random potential and magnetic field in each site

### DISCUSSION THREAD

**Discussion by Faridun22** [QuSpin/QuSpin#581] [Q&A] [answered]
Dear all,

How can I set a different potential and magnetic field in each site of the Fermi-Hubbard model?

Thank you in advance!

Faridun

**Comment by PFmix** [comment, marked as answer]
Assuming you are following e.g. [this example](http://weinbe58.github.io/QuSpin/examples/example13.html#example13-label),
it is relatively straightforward:
- for a random potential (I am assuming interaction potential) you change the line
```
interaction=[[U,i,i] for i in range(N_2d)]
```
into 
```
interaction=[[U*fct_random(),i,i] for i in range(N_2d)]
```
with fct_random() being the random generator of your choice
- for a random magnetic field, there is an extra step. Define e.g.
```
potential_r_u=[[-delta_mu*fct_random(),i] for i in range(N_2d)]
potential_r_d=[[-delta_mu*fct_random(),i] for i in range(N_2d)]

```
with delta_mu the amplitude of your fluctuations, assuming an average of zero for fct_random(). Then add to the static list of the Hamiltonian:
```
["n|",potential_r_u], ["|n",potential_r_d],
```
and you should have what you want.

**Reply by Faridun22** [reply]
Thank you so much for your reply,
One more question.  We consider Fermi-Hubbard model with additional disoder term (the ring with 20 sites). 
In QuSpin code, is there any possibility of conservation of the spin pattern in order to reduce the dimension of the Hilbert space?
We consider Neel spin pattern.

**Reply by PFmix** [reply]
I am not exactly sure I understood. Do you mean that you only want to consider states with, say, a zero magnetization?
If so, you can fix the number of fermion of spin up and spin down when generating the basis, and hence the total magnetization.

**Reply by Faridun22** [reply]
I am about symmetry of the system. There are two symmetries in our case:
1. The conservation of particle numbers.
2. The spin ordering symmetry.

The last one is guaranteed by the infinite on-site repulsion, which does not allow particles to exchange positions. This leads to a conserved spin pattern up to cyclic permutation. 
Here I present configurations which belong to the Nee’l spin pattern. 
<img width="293" alt="Снимок экрана 2022-11-04 в 19 44 32" src="https://user-images.githubusercontent.com/71867279/200030023-5e4ff61b-07d6-4d43-84a9-b24b15e6202b.png">
<img width="282" alt="Снимок экрана 2022-11-04 в 19 45 16" src="https://user-images.githubusercontent.com/71867279/200030171-1d944de5-823e-447b-af52-7a8d7a5989a1.png">

**Reply by PFmix** [reply]
The two pattern you show seem identical to me. But I thin I get what you mean: whatever configurations you consider needs to have this hidden Neel order. I believe it is possible to implement with user_basis... but I do not know how.
Alternatively, you can easily implement the conservation of number of particles of spin and down separately, and post select the states that verify the hidden order. Doing this method make you diagonalize a larger Hamiltonian which is naturally suboptimal.

**Reply by Faridun22** [reply]
Hello,

I would like to ask you a short question. 
Do you is there any way to create your own hamiltonian?
I would like to create FHM model with some disorder (with magnetic field and potential).
There is an example in QuSpin with chemical potential and onsite potential. But I need another one. Here is the disorder term which I wish:

<img width="477" alt="Снимок экрана 2022-11-21 в 08 39 26" src="https://user-images.githubusercontent.com/71867279/202973708-dd4eb2c0-862d-4949-a136-222817a98bc4.png">
